src/zenml/zen_server/exceptions.py

```diff
@@ -0,0 +1,175 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""REST API exception handling."""
+
+from typing import TYPE_CHECKING, Any, List, Optional, Tuple, Type
+
+import requests
+from pydantic import BaseModel
+from pydantic import ValidationError as PydanticValidationError
+
+from zenml.exceptions import (
+    AuthorizationException,
+    DoesNotExistException,
+    DuplicateRunNameError,
+    EntityExistsError,
+    IllegalOperationError,
+    SecretExistsError,
+    StackComponentExistsError,
+    StackExistsError,
+    ValidationError,
+    ZenKeyError,
+)
+
+if TYPE_CHECKING:
+    from fastapi import HTTPException
+
+
+class ErrorModel(BaseModel):
+    """Base class for error responses."""
+
+    detail: Any
+
+
+error_response = dict(model=ErrorModel)
+
+# Map exceptions to HTTP status codes. This is used in two ways:
+# 1. When an exception is raised in the server, the exception is converted to a
+#   error response with the appropriate status code and with the exception class
+#   name and arguments as the response body.
+# 2. When HTTP error response is received by the client, the error response is
+#  converted back to an exception based on the status code and the exception
+#  class name and arguments are extracted from the response body.
+#
+# NOTE: The order of the exceptions is important. The first matching
+# exception will be used, so the more specific exceptions should be
+# listed first. For error responses that cannot be unpacked into an
+# exception, the last (the more generic) exception in the list corresponding
+# to the status code will be used.
+REST_API_EXCEPTIONS: List[Tuple[Type[Exception], int]] = [
+    # 409 Conflict
+    (StackExistsError, 409),
+    (StackComponentExistsError, 409),
+    (SecretExistsError, 409),
+    (DuplicateRunNameError, 409),
+    (EntityExistsError, 409),
+    # 403 Forbidden
+    (IllegalOperationError, 403),
+    # 401 Unauthorized
+    (AuthorizationException, 401),
+    # 404 Not Found
+    (DoesNotExistException, 404),
+    (ZenKeyError, 404),
+    (KeyError, 404),
+    # 400 Bad Request
+    (PydanticValidationError, 400),
+    (ValidationError, 400),
+    (ValueError, 400),
+    # 422 Unprocessable Entity
+    (NotImplementedError, 422),
+    (PydanticValidationError, 422),
+    (ValueError, 400),
```

Comment:
even then it wouldn't make sense to have two mappings for the Value Error

Reply:
It depends. It could still make sense. The current implementation will use the last entry as default error when mapping back, i.e., both 400 and 422 would now be mapped back to `ValueError` if the correct exception class cannot be determined. I guess this was the purpose of that entry.

Reply:
This needs to be 422, good point, thanks for spotting it.